src/EditorFeatures/Core.Wpf/InlineErrors/InlineErrorAdornmentManager.cs

```diff
@@ -0,0 +1,216 @@
+                    var format = GetFormat(classificationType);
+                    tag.UpdateColor(format, element.Adornment);
+                }
+            }
+        }
+
+        private TextFormattingRunProperties GetFormat(IClassificationType classificationType)
+        {
+            return _formatMap.GetTextProperties(classificationType);
+        }
+
+        /// <summary>
+        /// Get the spans located on each line so that I can only display the first one that appears on the line
+        /// </summary>
+        private ImmutableDictionary<int, List<IMappingTagSpan<InlineErrorTag>>> GetSpansOnEachLine(NormalizedSnapshotSpanCollection changedSpanCollection)
+        {
+            _tagSpanToPointMap.Clear();
+            if (changedSpanCollection.IsEmpty())
+            {
+                return ImmutableDictionary<int, List<IMappingTagSpan<InlineErrorTag>>>.Empty;
+            }
+
+            var map = ImmutableDictionary.CreateBuilder<int, List<IMappingTagSpan<InlineErrorTag>>>();
+            var viewSnapshot = TextView.TextSnapshot;
+            var viewLines = TextView.TextViewLines;
+
+            foreach (var changedSpan in changedSpanCollection)
+            {
+                if (!viewLines.IntersectsBufferSpan(changedSpan))
+                {
+                    continue;
+                }
+
+                var tagSpans = TagAggregator.GetTags(changedSpan);
+                foreach (var tagMappingSpan in tagSpans)
+                {
+                    var point = tagMappingSpan.Span.Start.GetPoint(changedSpan.Snapshot, PositionAffinity.Predecessor);
+                    if (point == null)
+                    {
+                        continue;
+                    }
+
+                    var mappedPoint = TextView.BufferGraph.MapUpToSnapshot(
+                        point.Value, PointTrackingMode.Negative, PositionAffinity.Predecessor, TextView.VisualSnapshot);
+                    if (mappedPoint == null)
+                    {
+                        continue;
+                    }
+
+                    if (!TryMapToSingleSnapshotSpan(tagMappingSpan.Span, viewSnapshot, out var span))
+                    {
+                        continue;
+                    }
+
+                    if (!viewLines.IntersectsBufferSpan(span))
+                    {
+                        // span is outside of the view so we will not get geometry for it, but may 
+                        // spent a lot of time trying.
+                        continue;
+                    }
+
+                    var lineNum = mappedPoint.Value.GetContainingLine().LineNumber;
+                    if (!map.TryGetValue(lineNum, out var list))
+                    {
+                        list = new List<IMappingTagSpan<InlineErrorTag>>();
+                        map.Add(lineNum, list);
+                    }
+
+                    list.Add(tagMappingSpan);
+                    _tagSpanToPointMap.Add(tagMappingSpan, point.Value);
```

Comment:
that this mutates this state feels wonky.

Reply:
changhed